File: problems/0739-daily-temperatures/typescript/solution.ts
// Daily Temperatures — https://leetcode.com/problems/daily-temperatures/
export function dailyTemperatures(temperatures: number[]): number[] {
  const res = new Array<number>(temperatures.length).fill(0);
  const stack: number[] = []; // indices, temperatures decreasing
  for (let i = 0; i < temperatures.length; i++) {
    while (stack.length > 0 && temperatures[stack[stack.length - 1]] < temperatures[i]) {
      const j = stack[stack.length - 1];
      stack.pop();
      res[j] = i - j;
    }
    stack.push(i);
  }
  return res;
}

File: problems/0739-daily-temperatures/python/solution.py
def dailyTemperatures(temperatures):
    """Daily Temperatures — https://leetcode.com/problems/daily-temperatures/"""
    res = [0] * len(temperatures)
    stack = []  # indices, temperatures decreasing
    for i, temp in enumerate(temperatures):
        while stack and temperatures[stack[-1]] < temp:
            j = stack.pop()
            res[j] = i - j
        stack.append(i)
    return res
